File: flight_guard_ai/crypto.py
"""Cryptographic helper functions."""
from typing import Tuple

from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305


def encrypt(key: bytes, nonce: bytes, plaintext: bytes) -> Tuple[bytes, bytes]:
    """Encrypt data using ChaCha20-Poly1305.

    Returns a tuple of (ciphertext, tag) so callers can store the
    authentication tag separately from the ciphertext.
    """
    aead = ChaCha20Poly1305(key)
    ct = aead.encrypt(nonce, plaintext, b"")
    return ct[:-16], ct[-16:]


def decrypt(key: bytes, nonce: bytes, ciphertext: bytes, tag: bytes) -> bytes:
    """Decrypt data previously produced by :func:`encrypt`."""
    aead = ChaCha20Poly1305(key)
    return aead.decrypt(nonce, ciphertext + tag, b"")

File: firmware/src/crypto.rs
use chacha20poly1305::{aead::{AeadInPlace, KeyInit}, ChaCha20Poly1305, Key, Nonce};
use heapless::Vec;

/// Encrypt data in place using ChaCha20-Poly1305.
pub fn encrypt_in_place<const N: usize>(key: &[u8; 32], nonce: &[u8; 12], buf: &mut Vec<u8, N>) {
    let cipher = ChaCha20Poly1305::new(Key::from_slice(key));
    let nonce = Nonce::from_slice(nonce);
    let _ = cipher.encrypt_in_place(nonce, b"", buf);
}

/// Decrypt data in place.
pub fn decrypt_in_place<const N: usize>(key: &[u8; 32], nonce: &[u8; 12], buf: &mut Vec<u8, N>) -> bool {
    let cipher = ChaCha20Poly1305::new(Key::from_slice(key));
    let nonce = Nonce::from_slice(nonce);
    cipher.decrypt_in_place(nonce, b"", buf).is_ok()
}
